mbe: Implement `unsafe` attribute rules

joshtriplett opened the pull request:
This implements `unsafe attr` rules for declarative `macro_rules!` attributes, as specified in [RFC 3697](https://github.com/rust-lang/rfcs/pull/3697).

An invocation of an attribute that uses an `unsafe attr` rule requires the `unsafe(attr(...))` syntax.

An invocation of an attribute that uses an ordinary `attr` rule must *not* use the `unsafe(attr(...))` syntax.

`unsafe` is only supported on an `attr` rule, not any other kind of `macro_rules!` rule.

Tracking issue for `macro_rules!` attributes: https://github.com/rust-lang/rust/issues/143547

joshtriplett committed:
mbe: Rename a local variable to match corresponding field names

This simplifies subsequent initialization of enum variants.

joshtriplett committed:
mbe: Support `unsafe` attribute rules

joshtriplett committed:
mbe: Add parsing tests for `unsafe` macro rules

joshtriplett committed:
mbe: Add tests for `unsafe` attr invocation

mu001999 reviewed (approved):
LGTM

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146535/60a5372753460ef0db67c2307308a741b17771f3..a6bdf852e4117d70ba3f2b227396cc29fbbd28ff)*

vincenzopalazzo reviewed (commented):
While I was preparing a talk for the Rust 4linux summit, I did a first pass on this, and it looks good!

Will do a final pass on it next week

Thanks for bringing this forward

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146535/60a5372753460ef0db67c2307308a741b17771f3..a6bdf852e4117d70ba3f2b227396cc29fbbd28ff)*

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -706,6 +749,13 @@ pub fn compile_declarative_macro(
+                    unsafe_keyword_span,
+                    "`unsafe` is only supported on `attr` rules",
+                );
+                return dummy_syn_ext(err.emit());
```
Nit: no need to reset to dummy, the macro itself can still be used for recovery.
Same for fn-like rules.

joshtriplett replied:
Done.

joshtriplett committed:
mbe: `expand_invoc`: Add comment about not needing to check safety of `LegacyAttr` here

`LegacyAttr` is only used for builtin attributes, and builtin attributes
have their safety checked by `check_attribute_safety`, so we don't need
to check `unsafety` here.

petrochenkov commented:
r=me with the comments updated and review commits squashed.
@rustbot author
